src/services/createPropStore.ts

```diff
@@ -1,44 +1,63 @@
 import { InjectionKey, reactive } from 'vue'
 import { isWithComponent, isWithComponents } from '@/types/createRouteOptions'
+import { getPrefetchOption, PrefetchConfigs } from '@/types/prefetch'
 import { ResolvedRoute } from '@/types/resolved'
 import { Route } from '@/types/route'
-import { MaybePromise } from '@/types/utilities'
 
 export const propStoreKey: InjectionKey<PropStore> = Symbol()
 
 type ComponentProps = { id: string, name: string, props?: (params: Record<string, unknown>) => unknown }
 
 export type PropStore = {
+  prefetchProps: (route: ResolvedRoute, prefetch: PrefetchConfigs) => void,
   setProps: (route: ResolvedRoute) => void,
-  getProps: (id: string, name: string, params: unknown) => MaybePromise<unknown> | undefined,
+  getProps: (id: string, name: string, route: ResolvedRoute) => unknown,
 }
 
 export function createPropStore(): PropStore {
-  const store = reactive(new Map<string, unknown>())
-
-  function setProps(route: ResolvedRoute): void {
-    store.clear()
+  const store: Map<string, unknown> = reactive(new Map())
 
+  const prefetchProps: PropStore['prefetchProps'] = (route, prefetch) => {
     route.matches
-      .flatMap(match => getComponentProps(match))
+      .filter(match => getPrefetchOption({ ...prefetch, routePrefetch: match.prefetch }, 'props'))
+      .flatMap(getComponentProps)
       .forEach(({ id, name, props }) => {
         if (props) {
-          const key = getPropKey(id, name, route.params)
+          const key = getPropKey(id, name, route)
           const value = props(route.params)
 
-          store.set(key, value)
+          store.set(key, { prefetched: true, value })
         }
       })
   }
 
-  function getProps(id: string, name: string, params: unknown): MaybePromise<unknown> | undefined {
-    const key = getPropKey(id, name, params)
+  const setProps: PropStore['setProps'] = (route) => {
+    const componentProps = route.matches.flatMap(getComponentProps)
+    const routeKeys = componentProps.reduce<string[]>((routeKeys, { id, name, props }) => {
+      const key = getPropKey(id, name, route)
+
+      if (!props || store.has(key)) {
+        return routeKeys
+      }
+
+      const value = props(route.params)
+
+      store.set(key, value)
+
+      return [...routeKeys, key]
```

Comment:
super nit: I'd just push the new key and return the same array. Rather than creating a new array every time.